### PR TITLE
Add GHA check for formal spec SRP validity

### LINKED ISSUE #4549: Add a CI check for formal-ledger-spec SRP
Add a specific GitHub Action check just for this, so we can make sure that this SRP doesn't break `master`. :sweat_smile:

_Originally posted by @Lucsanszky in https://github.com/IntersectMBO/cardano-ledger/pull/4534#discussion_r1717413055_

### PR DESCRIPTION
**Pull request by Lucsanszky**
# Description
Resolves #4549 

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages.
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated.
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: .github/workflows/haskell.yml [hunk 1]
```diff
@@ -465,3 +465,34 @@ jobs:
...
+      run: |
+        RES=$(sed -ne \
+          '/ *location: *https:\/\/github.com\/IntersectMBO\/formal-ledger-specifications.git/,/--sha256/ p' \
+          cabal.project \
+        | sed -ne 's/^ *tag: *//p')
+        echo "TAG=$RES" >> $GITHUB_ENV
+
+    - uses: actions/checkout@v4
+      with:
+        repository: intersectmbo/formal-ledger-specifications
```
Comment (neilmayhew) [inline .github/workflows/haskell.yml]: You can avoid the checkout step and just use `git fetch` like this:

```
git fetch $LOCATION $TAG
```

This in turn means we could consider making this more general and have it apply to any SRP in `cabal.project`. However, maybe that should be a separate PR.

File: .github/workflows/haskell.yml [hunk 2]
```diff
@@ -465,3 +465,34 @@ jobs:
...
+
+    steps:
+    - uses: actions/checkout@v4
+    - name: Extract commit hash from formal-ledger-specifications SRP
+      run: |
+        RES=$(sed -ne \
+          '/ *location: *https:\/\/github.com\/IntersectMBO\/formal-ledger-specifications.git/,/--sha256/ p' \
+          cabal.project \
+        | sed -ne 's/^ *tag: *//p')
+        echo "TAG=$RES" >> $GITHUB_ENV
```
Comment (neilmayhew) [inline .github/workflows/haskell.yml]: Once you no longer need the second `checkout` action, I don't think you need to make these separate GHA steps. You could change `RES` to `TAG` and then use it lower down.

File: .github/workflows/haskell.yml (hunk 1 above)
Comment (Lucsanszky) [reply .github/workflows/haskell.yml]: Nice, thanks! 

The generalisation is a good idea but it should definitely be a separate PR and I think we can wait with that until the need arises.

File: .github/workflows/haskell.yml [hunk 3]
```diff
@@ -465,3 +465,27 @@ jobs:
...
+    - uses: actions/checkout@v4
+    - name: Check if provided formal-ledger-specifications SRP commit hash exists
+      run: |
+        TAG=$(sed -ne \
+          '/ *location: *https:\/\/github.com\/IntersectMBO\/formal-ledger-specifications.git/,/--sha256/ p' \
+          cabal.project \
+        | sed -ne 's/^ *tag: *//p')
+        # The unshallow option is required so we fetch
+        # the full history of the `MAlonzo-code` branch.
+        git fetch --unshallow https://github.com/intersectmbo/formal-ledger-specifications.git MAlonzo-code
```
Comment (neilmayhew) [inline .github/workflows/haskell.yml]: Maybe you already tested this, but I feel like the `--unshallow` isn't necessary because the fetch is coming from a separate repository that has no commits in common with this one, and so no commits will be suppressed.

It doesn't do any harm to leave it in there, of course, but if it's not needed then it does make things look more complicated than they need to be.

Comment (Lucsanszky) [reply .github/workflows/haskell.yml]: Interesting! Didn't know that. Indeed, it isn't required here, I've just checked it. Cheers!

**Review by neilmayhew — APPROVED** [review]
It's pleasingly lean and simple now! :smile:

**Comment by neilmayhew** [comment]
It's great to have this in our CI now
